torch/csrc/jit/passes/integer_value_refinement.cpp

```diff
@@ -0,0 +1,215 @@
+#include <ATen/core/jit_type.h>
+#include <jit/ir/ir.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/integer_value_refinement.h>
+#include <torch/csrc/jit/passes/value_refinement_utils.h>
+#include <torch/csrc/utils/memory.h>
+
+namespace torch {
+namespace jit {
+
+using IntegerRefinement = std::unordered_map<Value*, int64_t>;
+
+// see [value refinement algorithm] for full explanation.
+// When a comparison like `cond = x == 4` or `cond = x != 4` is made,
+// `cond` value carries information (refinements) about the value of `x`.
+// in an example like:
+// if x == 1:
+//    ...
+// we can substitute all uses of x dominated by the true block
+// with 1.
+
+struct IntegerValueRefiner {
+  IntegerValueRefiner(std::shared_ptr<Graph> graph)
+      : graph_(std::move(graph)) {}
+
+  bool run() {
+    if (!checkForPossibleRefinements(graph_->block())) {
+      return false;
+    }
+    IntegerRefinement refinements;
+    RefineIntegerValues(graph_->block(), refinements);
+    return changed_;
+  }
+
+  bool checkForPossibleRefinements(Block* b) {
+    for (Node* n : b->nodes()) {
+      if (n->matches("aten::eq(int a, int b) -> bool") ||
+          n->matches("aten::ne(int a, int b) -> bool")) {
+        for (size_t const_index : {0, 1}) {
+          auto non_const_index = 1 - const_index;
+          if (n->inputs().at(const_index)->node()->kind() == prim::Constant &&
+              n->inputs().at(non_const_index)->uses().size() > 1) {
+            return true;
+          }
+        }
+      }
+      for (Block* block : n->blocks()) {
+        if (checkForPossibleRefinements(block)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  void removeIfNodeOutputsWithRefinements(
+      Node* if_node,
+      IntegerRefinement& true_block_refinements,
+      IntegerRefinement& false_block_refinements) {
+    // we are looking for cases where we can replace
+    // both block outputs with the same value, which opens up
+    // further optimization opportunities
+    // The pass will already handle if both are refined to the same
+    // constant. Here, we add a case case where one block output
+    // is refined to a constant in the other block, and where the existing
+    // block output in the other block is the same constant
+    // x = 1
+    // if y == 1:
+    //    return x
+    // else:
+    //    return y
```

Comment:
It would be easier to understand code if you included TS IR dumps as well. It's a contiguous mental overhead to translate this into an IR in your head as you read the code (i.e. what's `block_output` in this example?)